tests/collect_nodes.cpp

```diff
@@ -0,0 +1,118 @@
+#include "ast/passes/collect_nodes.h"
+#include "gtest/gtest.h"
+
+#include <functional>
+#include <vector>
+
+namespace bpftrace::test::collect_nodes {
+
+using namespace bpftrace::ast;
+
+template <typename T>
+void test(const std::vector<std::reference_wrapper<T>> &expected,
+          const std::vector<std::reference_wrapper<T>> &actual)
+{
+  ASSERT_EQ(expected.size(), actual.size());
+  for (size_t i = 0; i < expected.size(); i++) {
+    EXPECT_EQ(&expected[i].get(), &actual[i].get());
+  }
+}
+
+TEST(CollectNodes, direct)
+{
+  auto &var = *new Variable{ "myvar", bpftrace::location{} };
```

Comment:
Why use `new` instead of the below?
```suggestion
  auto var = Variable("myvar", bpftrace::location{});
```
Also does GTest automatically cleanup the allocated memory?

Reply:
I wish I didn't have to - `new` is horrible :(

It's because the AST expects nodes to have been created on the heap and tries to delete them in the destructors: https://github.com/bpftrace/bpftrace/blob/d592be8fe65d7b49aa183eb1a05625df1743ed0c/src/ast/ast.cpp#L94

I started with the other approach but got crashes from "double frees" when `delete` was run on the objects.

Reply:
Ah, that's not too nice.